File: Sources/CursorStatusBar/CursorAPI.swift
import Foundation

enum APIError: Error, LocalizedError {
    case httpError(Int, String?)
    case noData
    case decodingError(String)

    var errorDescription: String? {
        switch self {
        case .httpError(let code, let body):
            return "HTTP \(code): \(body ?? "no body")"
        case .noData:
            return "No data received"
        case .decodingError(let msg):
            return "Decoding error: \(msg)"
        }
    }
}

actor CursorAPI {
    private let sessionToken: String
    private let userId: String

    init(sessionToken: String, userId: String) {
        self.sessionToken = sessionToken
        self.userId = userId
    }

    // MARK: - Fetch billing period start from legacy endpoint

    func fetchBillingPeriodStart() async throws -> Date {
        var components = URLComponents(string: "https://cursor.com/api/usage")!
        components.queryItems = [URLQueryItem(name: "user", value: userId)]

        var request = URLRequest(url: components.url!)
        request.httpMethod = "GET"
        request.setValue("WorkosCursorSessionToken=\(sessionToken)", forHTTPHeaderField: "Cookie")

        let (data, response) = try await URLSession.shared.data(for: request)
        try validateResponse(response, data: data)

        let legacy = try JSONDecoder().decode(LegacyUsageResponse.self, from: data)

        if let startStr = legacy.startOfMonth {
            let formatter = ISO8601DateFormatter()
            formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            if let date = formatter.date(from: startStr) {
                return date
            }
        }

        // Fallback: start of current calendar month
        let cal = Calendar.current
        return cal.date(from: cal.dateComponents([.year, .month], from: Date()))!
    }

    // MARK: - Fetch Usage Events (current API)

    func fetchUsageEvents(from startDate: Date, to endDate: Date, page: Int = 1, pageSize: Int = 1000) async throws -> [UsageEvent] {
        let url = URL(string: "https://cursor.com/api/dashboard/get-filtered-usage-events")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        applyHeaders(&request)

        let body: [String: Any] = [
            "teamId": 0,
            "startDate": String(Int(startDate.timeIntervalSince1970 * 1000)),
            "endDate": String(Int(endDate.timeIntervalSince1970 * 1000)),
            "page": page,
            "pageSize": pageSize
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        try validateResponse(response, data: data)

        let decoded = try JSONDecoder().decode(UsageEventsResponse.self, from: data)
        return decoded.usageEventsDisplay ?? []
    }

    // MARK: - Combined: fetch everything and aggregate

    func fetchDisplayData() async throws -> UsageDisplayData {
        // Get billing period start
        let billingStart = try await fetchBillingPeriodStart()
        let now = Date()
        let cal = Calendar.current

        // Fetch from the earlier of (billing start, 30 days ago)
        let thirtyDaysAgo = cal.date(byAdding: .day, value: -30, to: now)!
        let fetchStart = min(billingStart, thirtyDaysAgo)

        // Fetch all usage events
        let events = try await fetchUsageEvents(from: fetchStart, to: now)

        // Time boundaries
        let startOfToday = cal.startOfDay(for: now)
        let sevenDaysAgo = cal.date(byAdding: .day, value: -7, to: now)!

        // Aggregate by model (billing period) and by time buckets
        var byModel: [String: (count: Int, cents: Double, tokens: Int)] = [:]
        var totalCents: Double = 0
        var totalTokens: Int = 0

        var todayCents: Double = 0, todayReqs = 0, todayTokens = 0
        var week7Cents: Double = 0, week7Reqs = 0, week7Tokens = 0
        var days30Cents: Double = 0, days30Reqs = 0, days30Tokens = 0

        for event in events {
            let model = event.model ?? "unknown"
            let cents = event.costCents
            let tokens = event.tokenUsage?.totalTokens ?? 0

            // Parse timestamp (milliseconds since epoch)
            let timestampMs = Double(event.timestamp) ?? 0
            let eventDate = Date(timeIntervalSince1970: timestampMs / 1000.0)

            // Billing period totals (only events within billing period)
            if eventDate >= billingStart {
                totalCents += cents
                totalTokens += tokens

                var entry = byModel[model] ?? (count: 0, cents: 0, tokens: 0)
                entry.count += 1
                entry.cents += cents
                entry.tokens += tokens
                byModel[model] = entry
            }

            // Time bucket aggregation
            if eventDate >= startOfToday {
                todayCents += cents
                todayReqs += 1
                todayTokens += tokens
            }
            if eventDate >= sevenDaysAgo {
                week7Cents += cents
                week7Reqs += 1
                week7Tokens += tokens
            }
            if eventDate >= thirtyDaysAgo {
                days30Cents += cents
                days30Reqs += 1
                days30Tokens += tokens
            }
        }

        // Build line items sorted by cost descending
        let lineItems = byModel.map { (model, data) in
            UsageDisplayData.LineItem(
                modelName: model,
                requestCount: data.count,
                costDollars: data.cents / 100.0,
                totalTokens: data.tokens
            )
        }.sorted { $0.costDollars > $1.costDollars }

        let billingPeriodEventCount = byModel.values.reduce(0) { $0 + $1.count }

        return UsageDisplayData(
            totalRequests: billingPeriodEventCount,
            totalSpendDollars: totalCents / 100.0,
            totalTokens: totalTokens,
            lineItems: lineItems,
            billingPeriodStart: billingStart,
            today: PeriodSummary(
                label: "Today",
                requests: todayReqs,
                spendDollars: todayCents / 100.0,
                tokens: todayTokens
            ),
            last7Days: PeriodSummary(
                label: "Last 7 Days",
                requests: week7Reqs,
                spendDollars: week7Cents / 100.0,
                tokens: week7Tokens
            ),
            last30Days: PeriodSummary(
                label: "Last 30 Days",
                requests: days30Reqs,
                spendDollars: days30Cents / 100.0,
                tokens: days30Tokens
            )
        )
    }

    // MARK: - Private Helpers

    private func applyHeaders(_ request: inout URLRequest) {
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("WorkosCursorSessionToken=\(sessionToken)", forHTTPHeaderField: "Cookie")
        request.setValue("https://cursor.com", forHTTPHeaderField: "Origin")
        request.setValue("https://cursor.com/dashboard?tab=usage", forHTTPHeaderField: "Referer")
        request.setValue("same-origin", forHTTPHeaderField: "Sec-Fetch-Site")
        request.setValue("cors", forHTTPHeaderField: "Sec-Fetch-Mode")
        request.setValue("empty", forHTTPHeaderField: "Sec-Fetch-Dest")
        request.setValue("*/*", forHTTPHeaderField: "Accept")
        request.setValue("en", forHTTPHeaderField: "Accept-Language")
        request.setValue("no-cache", forHTTPHeaderField: "Cache-Control")
        request.setValue("no-cache", forHTTPHeaderField: "Pragma")
        request.setValue(
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
            forHTTPHeaderField: "User-Agent"
        )
    }

    private func validateResponse(_ response: URLResponse, data: Data) throws {
        guard let http = response as? HTTPURLResponse else { return }
        guard (200...299).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8)
            throw APIError.httpError(http.statusCode, body)
        }
    }
}

File: tauri/src-tauri/src/cursor_api.rs
use crate::models::*;
use chrono::{Datelike, DateTime, Duration, Local, NaiveDate, Utc};
use reqwest::Client;
use std::collections::HashMap;

pub struct CursorApi {
    client: Client,
    session_token: String,
    user_id: String,
}

impl CursorApi {
    pub fn new(session_token: String, user_id: String) -> Self {
        Self {
            client: Client::new(),
            session_token,
            user_id,
        }
    }

    /// Fetch the billing period start date from the legacy endpoint.
    async fn fetch_billing_period_start(&self) -> Result<DateTime<Utc>, Box<dyn std::error::Error + Send + Sync>> {
        let url = format!("https://cursor.com/api/usage?user={}", self.user_id);

        let resp = self
            .client
            .get(&url)
            .header("Cookie", format!("WorkosCursorSessionToken={}", self.session_token))
            .send()
            .await?;

        let status = resp.status();
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            return Err(format!("HTTP {}: {}", status, body).into());
        }

        let legacy: LegacyUsageResponse = resp.json().await?;

        if let Some(start_str) = legacy.start_of_month() {
            if let Ok(dt) = DateTime::parse_from_rfc3339(start_str) {
                return Ok(dt.with_timezone(&Utc));
            }
            // Try parsing as ISO 8601 with fractional seconds
            if let Ok(dt) = chrono::DateTime::parse_from_str(start_str, "%Y-%m-%dT%H:%M:%S%.fZ") {
                return Ok(dt.with_timezone(&Utc));
            }
        }

        // Fallback: start of current calendar month
        let now = Local::now();
        let start = NaiveDate::from_ymd_opt(now.year(), now.month(), 1)
            .unwrap()
            .and_hms_opt(0, 0, 0)
            .unwrap();
        Ok(DateTime::from_naive_utc_and_offset(start, Utc))
    }

    /// Fetch usage events from the current API.
    async fn fetch_usage_events(
        &self,
        from: DateTime<Utc>,
        to: DateTime<Utc>,
    ) -> Result<Vec<UsageEvent>, Box<dyn std::error::Error + Send + Sync>> {
        let url = "https://cursor.com/api/dashboard/get-filtered-usage-events";

        let body = serde_json::json!({
            "teamId": 0,
            "startDate": from.timestamp_millis().to_string(),
            "endDate": to.timestamp_millis().to_string(),
            "page": 1,
            "pageSize": 1000
        });

        let resp = self
            .client
            .post(url)
            .header("Content-Type", "application/json")
            .header("Cookie", format!("WorkosCursorSessionToken={}", self.session_token))
            .header("Origin", "https://cursor.com")
            .header("Referer", "https://cursor.com/dashboard?tab=usage")
            .header("Sec-Fetch-Site", "same-origin")
            .header("Sec-Fetch-Mode", "cors")
            .header("Sec-Fetch-Dest", "empty")
            .header("Accept", "*/*")
            .header("Accept-Language", "en")
            .header("Cache-Control", "no-cache")
            .header("Pragma", "no-cache")
            .header(
                "User-Agent",
                "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
            )
            .body(body.to_string())
            .send()
            .await?;

        let status = resp.status();
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            return Err(format!("HTTP {}: {}", status, body).into());
        }

        let events_resp: UsageEventsResponse = resp.json().await?;
        Ok(events_resp.usage_events_display.unwrap_or_default())
    }

    /// Fetch all data and aggregate into display format.
    pub async fn fetch_display_data(&self) -> Result<UsageDisplayData, Box<dyn std::error::Error + Send + Sync>> {
        let billing_start = self.fetch_billing_period_start().await?;
        let now = Utc::now();

        // Fetch from the earlier of (billing start, 30 days ago)
        let thirty_days_ago = now - Duration::days(30);
        let fetch_start = billing_start.min(thirty_days_ago);

        let events = self.fetch_usage_events(fetch_start, now).await?;

        // Time boundaries
        let start_of_today = Local::now()
            .date_naive()
            .and_hms_opt(0, 0, 0)
            .unwrap()
            .and_local_timezone(Local)
            .unwrap()
            .with_timezone(&Utc);
        let seven_days_ago = now - Duration::days(7);

        // Aggregate by model (billing period) and by time buckets
        let mut by_model: HashMap<String, (i32, f64, i64)> = HashMap::new();
        let mut total_cents: f64 = 0.0;
        let mut total_tokens: i64 = 0;

        let mut today_cents: f64 = 0.0;
        let mut today_reqs: i32 = 0;
        let mut today_tokens: i64 = 0;

        let mut week7_cents: f64 = 0.0;
        let mut week7_reqs: i32 = 0;
        let mut week7_tokens: i64 = 0;

        let mut days30_cents: f64 = 0.0;
        let mut days30_reqs: i32 = 0;
        let mut days30_tokens: i64 = 0;

        for event in &events {
            let model = event.model.clone().unwrap_or_else(|| "unknown".to_string());
            let cents = event.cost_cents();
            let tokens = event
                .token_usage
                .as_ref()
                .map(|t| t.total_tokens())
                .unwrap_or(0);

            // Parse timestamp (milliseconds since epoch)
            let timestamp_ms: f64 = event.timestamp.parse().unwrap_or(0.0);
            let event_date = DateTime::from_timestamp_millis(timestamp_ms as i64)
                .unwrap_or(DateTime::UNIX_EPOCH);

            // Billing period totals
            if event_date >= billing_start {
                total_cents += cents;
                total_tokens += tokens;

                let entry = by_model.entry(model).or_insert((0, 0.0, 0));
                entry.0 += 1;
                entry.1 += cents;
                entry.2 += tokens;
            }

            // Time bucket aggregation
            if event_date >= start_of_today {
                today_cents += cents;
                today_reqs += 1;
                today_tokens += tokens;
            }
            if event_date >= seven_days_ago {
                week7_cents += cents;
                week7_reqs += 1;
                week7_tokens += tokens;
            }
            if event_date >= thirty_days_ago {
                days30_cents += cents;
                days30_reqs += 1;
                days30_tokens += tokens;
            }
        }

        // Build line items sorted by cost descending
        let mut line_items: Vec<LineItem> = by_model
            .into_iter()
            .map(|(model, (count, cents, tokens))| LineItem {
                model_name: model,
                request_count: count,
                cost_dollars: cents / 100.0,
                total_tokens: tokens,
            })
            .collect();
        line_items.sort_by(|a, b| b.cost_dollars.partial_cmp(&a.cost_dollars).unwrap());

        let billing_period_event_count: i32 = line_items.iter().map(|i| i.request_count).sum();

        Ok(UsageDisplayData {
            total_requests: billing_period_event_count,
            total_spend_dollars: total_cents / 100.0,
            total_tokens,
            line_items,
            billing_period_start: billing_start.to_rfc3339(),
            today: PeriodSummary {
                label: "Today".to_string(),
                requests: today_reqs,
                spend_dollars: today_cents / 100.0,
                tokens: today_tokens,
            },
            last7_days: PeriodSummary {
                label: "Last 7 Days".to_string(),
                requests: week7_reqs,
                spend_dollars: week7_cents / 100.0,
                tokens: week7_tokens,
            },
            last30_days: PeriodSummary {
                label: "Last 30 Days".to_string(),
                requests: days30_reqs,
                spend_dollars: days30_cents / 100.0,
                tokens: days30_tokens,
            },
        })
    }
}
